src/backend/distributed/executor/executor.c

```diff
@@ -1301,6 +1331,80 @@ ManageWorkerPool(WorkerPool *workerPool)
 }
 
 
+/*
+ * CheckConnectionTimeout makes sure that the execution enforces the connection
+ * establishment timeout defined by the user (NodeConnectionTimeout).
+ *
+ * The rule is that if a worker pool has already ready tasks but no active connections
+ * at all, take an action. For the types of actions, see the comments in the function.
+ *
+ * Enforcing the timeout per pool (over per session) helps the execution to continue
+ * even if we can establish a single connection as we expect to have target pool size
+ * number of connections. In the end, the executor is capable of using one connection
+ * to execute multiple tasks.
+ */
+static void
+CheckConnectionTimeout(WorkerPool *workerPool)
+{
+	DistributedExecution *execution = workerPool->distributedExecution;
+	TimestampTz connectionStart = workerPool->connectionStart;
+	TimestampTz now = GetCurrentTimestamp();
+
+	int activeConnectionCount = workerPool->activeConnectionCount;
+	int readyTaskCount = workerPool->readyTaskCount;
+
+	int requiredActiveConnectionCount = 1;
+	bool tasksReadyToExecute = readyTaskCount > 0;
+
+	if (list_length(workerPool->sessionList) == 0)
+	{
+		/* no connection has been planned for the pool yet */
+		Assert(connectionStart == 0);
+		return;
+	}
+
+	/*
+	 * This is a special case where we assign tasks to sessions even before
+	 * the connections are established. So, make sure to apply similar
+	 * restrictions. In this case, make sure that we get all the connections
+	 * established.
+	 */
+	if (ForceMaxQueryParallelization)
+	{
+		requiredActiveConnectionCount = list_length(workerPool->sessionList);
```

Comment:
I'm kind of wondering whether it would be cleaner to distinguish between critical and non-critical connections, but it would probably be more trouble than it's worth. I think the long-term solution is to remove the flag.